src/server/pkg/hashtree/hashtree.go

```diff
@@ -256,9 +277,12 @@ func (h *hashtree) PutFile(path string, blockRefs []*pfs.BlockRef) error {
 		}
 		insertStr(&node.DirNode.Children, child)
 		node.SubtreeSize += sizeGrowth
-		h.updateHash(parent)
+		h.changed[parent] = true
 		return nil
-	})
+	}); err != nil {
+		return err
+	}
+	return h.canonicalize("/")
```

Comment:
Might be missing something obvious, but if you call `h.canonicalize("/")` in every PutFile, aren't you hashing the entire tree every time?